File: dotnet/Fingerprint.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;

namespace Shear
{
    public class PartialFingerprint
    {
        public long Length;

        public PartialFingerprint(FileInfo file)
        {
            Length = file.Length;
        }

        public override bool Equals(object obj)
        {
            var other = obj as PartialFingerprint;
            if (other == null)
                return false;

            return Length == other.Length;
        }

        public override int GetHashCode()
        {
            return Length.GetHashCode();
        }
    }

    public class Fingerprint
    {
        private const int MaxSize = 0x800000;

        public long Length;

        byte[] _head;
        byte[] _tail;

        public IEnumerable<byte> Head => _head;
        public IEnumerable<byte> Tail => _tail == null
            ? Enumerable.Empty<byte>()
            : _tail;

        public Fingerprint(string path)
        {
            var file = File.OpenRead(path);

            Length = file.Length;
            _head = Hash(file, file.Length > MaxSize ? MaxSize : (int)file.Length);
            _tail = null as byte[];

            if (Length > MaxSize)
            {
                var tailLength = MaxSize > Length
                    ? Math.Min(MaxSize, (int)Length - MaxSize)
                    : MaxSize;

                file.Seek(tailLength, SeekOrigin.End);
                _tail = Hash(file, tailLength);
            }
        }

        public override bool Equals(object obj)
        {
            var other = obj as Fingerprint;
            if (other == null)
                return false;

            return Length == other.Length
                && Head.SequenceEqual(other.Head)
                && Tail.SequenceEqual(other.Tail);
        }

        public override int GetHashCode()
        {
            var hash = Length.GetHashCode();
            hash = hash * 17 + HashArray(_head);

            if (_tail != null)
            {
                hash = hash * 17 + HashArray(_tail);
            }

            return hash;
        }

        private static byte[] Hash(Stream stream, int size)
        {
            var hasher = SHA256.Create();
            var buffer = new byte[size];

            var bytesRead = stream.Read(buffer, 0, size);

            hasher.ComputeHash(buffer, 0, bytesRead);
            return hasher.Hash;
        }

        private static int HashArray(byte[] array)
        {
            int hash = 17;
            foreach (var element in array)
            {
                hash = hash * 31 + element;
            }
            return hash;
        }
    }
}

File: src/fingerprint.rs
use std::io::{self, Read};
use std::path::Path;

const MAX_SIZE: u64 = 0x0080_0000;

#[derive(Debug, Eq, PartialEq, Hash)]
pub struct PartialFingerprint(u64);

impl PartialFingerprint {
    pub fn from_path<T: AsRef<Path>>(path: T) -> io::Result<Self> {
        use std::fs::File;
        let file = File::open(path)?;
        let meta = file.metadata()?;
        Ok(PartialFingerprint(meta.len()))
    }
}

#[derive(Debug, Eq, PartialEq, Hash)]
pub struct Fingerprint {
    length: u64,
    head: Vec<u8>, // Here we use a vec solely because an array lacks eq and hash definitions.
    tail: Option<Vec<u8>>,
}

impl Fingerprint {
    pub fn from_path<T: AsRef<Path>>(path: T) -> io::Result<Self> {
        use std::cmp;
        use std::fs::File;
        use std::io::Seek;
        use std::io::SeekFrom;

        let mut file = File::open(path)?;
        let meta = file.metadata()?;

        let head = hash(&mut file)?;
        let tail = {
            if meta.len() > MAX_SIZE {
                let extra = if MAX_SIZE > meta.len() {
                    cmp::min(MAX_SIZE, meta.len() - MAX_SIZE)
                } else {
                    MAX_SIZE
                };

                file.seek(SeekFrom::End(extra as i64))?;
                Some(hash(&mut file)?)
            } else {
                None
            }
        };

        Ok(Fingerprint {
            length: meta.len(),
            head,
            tail,
        })
    }
}

fn hash(mut stream: impl Read) -> io::Result<Vec<u8>> {
    use sha2::Sha256;
    use digest::{FixedOutput, Input};

    // We need box emplacement here because, otherwise, it blows the stack.
    let mut buf = box [0u8; MAX_SIZE as usize];
    let len = stream.read(&mut *buf)?;

    let mut hasher = Sha256::default();
    hasher.process(&buf[..len]);

    Ok(hasher.fixed_result().into_iter().collect())
}
